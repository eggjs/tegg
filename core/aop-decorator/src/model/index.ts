export * from './Aspect.js';
export * from './PointcutInfo.js';
