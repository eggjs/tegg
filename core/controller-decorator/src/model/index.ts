export * from './types';

export * from './MethodMeta';
export * from './ControllerMetadata';
export * from './HTTPMethodMeta';
export * from './HTTPControllerMeta';
export * from './HTTPRequest';
export * from './HTTPResponse';
