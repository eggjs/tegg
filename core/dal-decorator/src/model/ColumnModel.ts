import snakecase from 'lodash.snakecase';
import type { ColumnFormat, ColumnParams, ColumnTypeParams } from '@eggjs/tegg-types';

export class ColumnModel {
  columnName: string;
  propertyName: string;
  type: ColumnTypeParams;
  canNull: boolean;
  default?: string;
  comment?: string;
  visible?: boolean;
  autoIncrement?: boolean;
  uniqueKey?: boolean;
  primaryKey?: boolean;
  collate?: string;
  columnFormat?: ColumnFormat;
  engineAttribute?: string;
  secondaryEngineAttribute?: string;

  constructor(params: {
    columnName: string;
    propertyName: string;
    type: ColumnTypeParams;
    canNull: boolean;
    default?: string;
    comment?: string;
    visible?: boolean;
    autoIncrement?: boolean;
    uniqueKey?: boolean;
    primaryKey?: boolean;
    collate?: string;
    columnFormat?: ColumnFormat;
    engineAttribute?: string;
    secondaryEngineAttribute?: string;
  }) {
    this.columnName = params.columnName;
    this.propertyName = params.propertyName;
    this.type = params.type;
    this.canNull = params.canNull;
    this.default = params.default;
    this.comment = params.comment;
    this.visible = params.visible;
    this.autoIncrement = params.autoIncrement;
    this.uniqueKey = params.uniqueKey;
    this.primaryKey = params.primaryKey;
    this.collate = params.collate;
    this.columnFormat = params.columnFormat;
    this.engineAttribute = params.engineAttribute;
    this.secondaryEngineAttribute = params.secondaryEngineAttribute;
  }

  static build(property: string, type: ColumnTypeParams, params?: ColumnParams) {
    const columnName = params?.name ?? snakecase(property);
    let canNull = params?.canNull ?? false;
    if (params?.primaryKey) {
      canNull = false;
    }
    return new ColumnModel({
      columnName,
      propertyName: property,
      type,
      canNull,
      default: params?.default,
      comment: params?.comment,
      visible: params?.visible,
      autoIncrement: params?.autoIncrement,
      uniqueKey: params?.uniqueKey,
      primaryKey: params?.primaryKey,
      collate: params?.collate,
      columnFormat: params?.columnFormat,
      engineAttribute: params?.engineAttribute,
      secondaryEngineAttribute: params?.secondaryEngineAttribute,
    });
  }
}
