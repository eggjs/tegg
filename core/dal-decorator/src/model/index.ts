export * from './ColumnModel.js';
export * from './IndexModel.js';
export * from './TableModel.js';
