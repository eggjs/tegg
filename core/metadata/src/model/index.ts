export * from './graph/index.js';
export * from './ProtoDescriptor/index.js';
export * from './AppGraph.js';
export * from './EggPrototype.js';
export * from './LoadUnit.js';
export * from './ModuleDescriptor.js';
export * from './ProtoDescriptorHelper.js';
