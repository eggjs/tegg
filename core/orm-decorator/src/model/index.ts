export * from './AttributeMeta.js';
export * from './IndexMeta.js';
export * from './ModelMetadata.js';
