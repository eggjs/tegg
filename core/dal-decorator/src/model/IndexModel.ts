import { EggProtoImplClass, IndexType } from '@eggjs/tegg-types';
import type { IndexParams, IndexStoreType } from '@eggjs/tegg-types';
import { ColumnModel } from './ColumnModel';

export interface IndexKey {
  columnName: string;
  propertyName: string;
}

export class IndexModel {
  name: string;
  keys: IndexKey[];
  type: IndexType;

  storeType?: IndexStoreType;
  comment?: string;
  engineAttribute?: string;
  secondaryEngineAttribute?: string;
  parser?: string;

  constructor(params: {
    name: string;
    keys: IndexKey[];
    type: IndexType;
    storeType?: IndexStoreType;
    comment?: string;
    engineAttribute?: string;
    secondaryEngineAttribute?: string;
    parser?: string;
  }) {
    this.name = params.name;
    this.keys = params.keys;
    this.type = params.type;
    this.storeType = params.storeType;
    this.comment = params.comment;
    this.engineAttribute = params.engineAttribute;
    this.secondaryEngineAttribute = params.secondaryEngineAttribute;
    this.parser = params.parser;
  }

  static buildIndexName(keys: string[], type: IndexType) {
    const prefix = type === IndexType.UNIQUE ? 'uk_' : 'idx_';
    return prefix + keys.join('_');
  }

  static build(params: IndexParams, columns: ColumnModel[], clazz: EggProtoImplClass<unknown>) {
    const type = params.type ?? IndexType.INDEX;
    const keys: Array<IndexKey> = params.keys.map(t => {
      const column = columns.find(c => c.propertyName === t);
      if (!column) {
        throw new Error(`Table ${clazz.name} index configuration error: has no property named "${t}"`);
      }
      return {
        propertyName: column!.propertyName,
        columnName: column!.columnName,
      };
    });
    const name = params.name ?? IndexModel.buildIndexName(keys.map(t => t.columnName), type);
    return new IndexModel({
      name,
      keys,
      type,
      storeType: params.storeType,
      comment: params.comment,
      engineAttribute: params.engineAttribute,
      secondaryEngineAttribute: params.secondaryEngineAttribute,
      parser: params.parser,
    });
  }
}
