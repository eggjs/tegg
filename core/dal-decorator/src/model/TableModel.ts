import assert from 'node:assert';
import pluralize from 'pluralize';
import snakecase from 'lodash.snakecase';
import { IndexType } from '@eggjs/tegg-types';
import type { CompressionType, EggProtoImplClass, InsertMethod, RowFormat } from '@eggjs/tegg-types';
import { ColumnModel } from './ColumnModel';
import { IndexModel } from './IndexModel';
import { TableInfoUtil } from '../util/TableInfoUtil';
import { ColumnInfoUtil } from '../util/ColumnInfoUtil';
import { IndexInfoUtil } from '../util/IndexInfoUtil';

export class TableModel<T=object> {
  clazz: EggProtoImplClass<T>;
  name: string;
  columns: Array<ColumnModel>;
  indices: Array<IndexModel>;
  dataSourceName: string;
  comment?: string;
  autoExtendSize?: number;
  autoIncrement?: number;
  avgRowLength?: number;
  characterSet?: string;
  collate?: string;
  compression?: CompressionType;
  encryption?: boolean;
  engine?: string;
  engineAttribute?: string;
  insertMethod?: InsertMethod;
  keyBlockSize?: number;
  maxRows?: number;
  minRows?: number;
  rowFormat?: RowFormat;
  secondaryEngineAttribute?: string;

  constructor(params: {
    clazz: EggProtoImplClass<T>;
    name: string;
    dataSourceName: string;
    columns: Array<ColumnModel>;
    indices: Array<IndexModel>;
    comment?: string;
    autoExtendSize?: number;
    autoIncrement?: number;
    avgRowLength?: number;
    characterSet?: string;
    collate?: string;
    compression?: CompressionType;
    encryption?: boolean;
    engine?: string;
    engineAttribute?: string;
    insertMethod?: InsertMethod;
    keyBlockSize?: number;
    maxRows?: number;
    minRows?: number;
    rowFormat?: RowFormat;
    secondaryEngineAttribute?: string;
  }) {
    this.clazz = params.clazz;
    this.name = params.name;
    this.dataSourceName = params.dataSourceName;
    this.columns = params.columns;
    this.indices = params.indices;
    this.comment = params.comment;
    this.autoExtendSize = params.autoExtendSize;
    this.autoIncrement = params.autoIncrement;
    this.avgRowLength = params.avgRowLength;
    this.characterSet = params.characterSet;
    this.collate = params.collate;
    this.compression = params.compression;
    this.encryption = params.encryption;
    this.engine = params.engine;
    this.engineAttribute = params.engineAttribute;
    this.insertMethod = params.insertMethod;
    this.keyBlockSize = params.keyBlockSize;
    this.maxRows = params.maxRows;
    this.minRows = params.minRows;
    this.rowFormat = params.rowFormat;
    this.secondaryEngineAttribute = params.secondaryEngineAttribute;
  }

  getPrimary(): IndexModel | undefined {
    const index = this.indices.find(t => t.type === IndexType.PRIMARY);
    if (index) {
      return index;
    }
    const primaryColumn = this.columns.filter(t => t.primaryKey === true);
    return new IndexModel({
      name: 'PRIMARY',
      type: IndexType.PRIMARY,
      keys: primaryColumn.map(t => {
        return {
          columnName: t.columnName,
          propertyName: t.propertyName,
        };
      }),
    });
  }

  static build<T>(clazz: EggProtoImplClass<T>): TableModel<T> {
    const params = TableInfoUtil.getTableParams(clazz as EggProtoImplClass);
    const name = params?.name ?? snakecase(pluralize(clazz.name));
    const columnInfoMap = ColumnInfoUtil.getColumnInfoMap(clazz as EggProtoImplClass);
    const columnTypeMap = ColumnInfoUtil.getColumnTypeMap(clazz as EggProtoImplClass);
    const dataSourceName = params?.dataSourceName ?? 'default';
    assert(TableInfoUtil.getIsTable(clazz as EggProtoImplClass), `${name} is not Table`);
    assert(columnTypeMap, `${name} has no columns`);
    const columns: Array<ColumnModel> = [];
    const indices: Array<IndexModel> = [];
    for (const [ property, columnType ] of columnTypeMap?.entries()) {
      const columnParam = columnInfoMap?.get(property);
      columns.push(ColumnModel.build(property, columnType, columnParam));
    }

    const indexList = IndexInfoUtil.getIndexList(clazz as EggProtoImplClass);
    for (const index of indexList) {
      indices.push(IndexModel.build(index, columns));
    }

    return new TableModel({
      clazz,
      name,
      columns,
      indices,
      dataSourceName,

      comment: params?.comment,
      autoExtendSize: params?.autoExtendSize,
      autoIncrement: params?.autoIncrement,
      avgRowLength: params?.avgRowLength,
      characterSet: params?.characterSet,
      collate: params?.collate,
      compression: params?.compression,
      encryption: params?.encryption,
      engine: params?.engine,
      engineAttribute: params?.engineAttribute,
      insertMethod: params?.insertMethod,
      keyBlockSize: params?.keyBlockSize,
      maxRows: params?.maxRows,
      minRows: params?.minRows,
      rowFormat: params?.rowFormat,
      secondaryEngineAttribute: params?.secondaryEngineAttribute,
    });
  }
}
