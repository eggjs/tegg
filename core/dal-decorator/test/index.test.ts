import assert from 'node:assert';
import { Foo } from './fixtures/modules/dal/Foo';
import { ColumnInfoUtil, ColumnType, IndexInfoUtil, IndexType, TableInfoUtil } from '..';
import { TableModel } from '../src/model/TableModel';

describe('test/dal/index.test.ts', () => {
  it('decorator should work', () => {
    const columnInfoMap = ColumnInfoUtil.getColumnInfoMap(Foo);
    const columnTypeMap = ColumnInfoUtil.getColumnTypeMap(Foo);

    const indexList = IndexInfoUtil.getIndexList(Foo);

    const tableInfo = TableInfoUtil.getTableParams(Foo);
    const isTable = TableInfoUtil.getIsTable(Foo);

    assert.deepStrictEqual(columnInfoMap, new Map([
      [
        'id',
        {
          primaryKey: true,
        },
      ],
    ]));
    assert.deepStrictEqual(columnTypeMap, new Map([
      [
        'id',
        {
          type: ColumnType.INT,
        },
      ], [
        'name',
        {
          type: ColumnType.VARCHAR,
          length: 100,
        },
      ],
    ]));

    assert.deepStrictEqual(indexList, [{
      keys: [ 'name' ],
      type: IndexType.UNIQUE,
    }]);

    assert.deepStrictEqual(tableInfo, {
      comment: 'foo table',
    });

    assert.equal(isTable, true);
  });

  it('model should work', () => {
    const table = TableModel.build(Foo);
    assert(table);
    assert(table.name === 'foos');
    assert.equal(table.columns.length, 2);
    assert.equal(table.indices.length, 1);
  });
});
