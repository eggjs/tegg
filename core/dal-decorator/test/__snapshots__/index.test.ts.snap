// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test/dal/index.test.ts > should export stable 1`] = `
{
  "Column": [Function],
  "ColumnFormat": {
    "DEFAULT": "DEFAULT",
    "DYNAMIC": "DYNAMIC",
    "FIXED": "FIXED",
  },
  "ColumnInfoUtil": [Function],
  "ColumnModel": [Function],
  "ColumnType": {
    "BIGINT": "BIGINT",
    "BINARY": "BINARY",
    "BIT": "BIT",
    "BLOB": "BLOB",
    "BOOL": "BOOL",
    "CHAR": "CHAR",
    "DATE": "DATE",
    "DATETIME": "DATETIME",
    "DECIMAL": "DECIMAL",
    "DOUBLE": "DOUBLE",
    "ENUM": "ENUM",
    "FLOAT": "FLOAT",
    "GEOMETRY": "GEOMETRY",
    "GEOMETRYCOLLECTION": "GEOMETRYCOLLECTION",
    "INT": "INT",
    "JSON": "JSON",
    "LINESTRING": "LINESTRING",
    "LONGBLOB": "LONGBLOB",
    "LONGTEXT": "LONGTEXT",
    "MEDIUMBLOB": "MEDIUMBLOB",
    "MEDIUMINT": "MEDIUMINT",
    "MEDIUMTEXT": "MEDIUMTEXT",
    "MULTILINESTRING": "MULTILINESTRING",
    "MULTIPOINT": "MULTIPOINT",
    "MULTIPOLYGON": "MULTIPOLYGON",
    "POINT": "POINT",
    "POLYGON": "POLYGON",
    "SET": "SET",
    "SMALLINT": "SMALLINT",
    "TEXT": "TEXT",
    "TIME": "TIME",
    "TIMESTAMP": "TIMESTAMP",
    "TINYBLOB": "TINYBLOB",
    "TINYINT": "TINYINT",
    "TINYTEXT": "TINYTEXT",
    "VARBINARY": "VARBINARY",
    "VARCHAR": "VARCHAR",
    "YEAR": "YEAR",
  },
  "CompressionType": {
    "LZ4": "LZ4",
    "NONE": "NONE",
    "ZLIB": "ZLIB",
  },
  "DAL_COLUMN_INFO_MAP": Symbol(EggPrototype#dalColumnInfoMap),
  "DAL_COLUMN_TYPE_MAP": Symbol(EggPrototype#dalColumnTypeMap),
  "DAL_INDEX_LIST": Symbol(EggPrototype#dalIndexList),
  "DAL_IS_DAO": Symbol(EggPrototype#dalIsDao),
  "DAL_IS_TABLE": Symbol(EggPrototype#dalIsTable),
  "DAL_TABLE_PARAMS": Symbol(EggPrototype#dalTableParams),
  "Dao": [Function],
  "DaoInfoUtil": [Function],
  "DataSourceInjectName": "dataSource",
  "DataSourceQualifier": [Function],
  "DataSourceQualifierAttribute": Symbol(Qualifier.DataSource),
  "DeleteResult": undefined,
  "Index": [Function],
  "IndexInfoUtil": [Function],
  "IndexModel": [Function],
  "IndexStoreType": {
    "BTREE": "BTREE",
    "HASH": "HASH",
  },
  "IndexType": {
    "FULLTEXT": "FULLTEXT",
    "INDEX": "INDEX",
    "PRIMARY": "PRIMARY",
    "SPATIAL": "SPATIAL",
    "UNIQUE": "UNIQUE",
  },
  "InsertMethod": {
    "FIRST": "FIRST",
    "LAST": "LAST",
    "NO": "NO",
  },
  "InsertResult": undefined,
  "RowFormat": {
    "COMPACT": "COMPACT",
    "COMPRESSED": "COMPRESSED",
    "DEFAULT": "DEFAULT",
    "DYNAMIC": "DYNAMIC",
    "FIXED": "FIXED",
    "REDUNDANT": "REDUNDANT",
  },
  "SpatialHelper": [Function],
  "SqlType": {
    "BLOCK": "BLOCK",
    "DELETE": "DELETE",
    "INSERT": "INSERT",
    "SELECT": "SELECT",
    "UPDATE": "UPDATE",
  },
  "TABLE_CLAZZ_LIST": [
    [Function],
  ],
  "Table": [Function],
  "TableInfoUtil": [Function],
  "TableModel": [Function],
  "Templates": {
    "BASE_DAO": "base_dao",
    "DAO": "dao",
    "EXTENSION": "extension",
  },
  "UpdateResult": undefined,
}
`;
